src/Surfnet/StepupSelfService/SelfServiceBundle/Controller/SelfVetController.php

```diff
@@ -0,0 +1,209 @@
+        $candidateSecondFactorLoa = $this->loaResolutionService->getLoaByLevel($candidateSecondFactorLoa);
+
+        $this->logger->notice(
+            sprintf(
+                'Creating AuthNRequest requiring a LoA %s or higher token for self vetting.',
+                $candidateSecondFactorLoa
+            )
+        );
+        $authenticationRequest = $this->authenticationRequestFactory->createSecondFactorTestRequest(
+            $identity->nameId,
+            $candidateSecondFactorLoa
+        );
+
+        $this->session->set(
+            self::SELF_VET_SESSION_ID,
+            new SelfVetRequestId($authenticationRequest->getRequestId(), $secondFactorId)
+        );
+
+        $samlLogger = $this->samlLogger->forAuthentication($authenticationRequest->getRequestId());
+        $samlLogger->notice('Sending authentication request to the second factor only IdP');
+
+        return $this->redirectBinding->createRedirectResponseFor($authenticationRequest);
+    }
+
+    public function consumeSelfVetAssertionAction(Request $httpRequest, string $secondFactorId)
+    {
+        if (!$this->session->has(self::SELF_VET_SESSION_ID)) {
+            $this->logger->error(
+                'Received an authentication response for self vetting a second factor, but no response was expected'
+            );
+            throw new AccessDeniedHttpException('Did not expect an authentication response');
+        }
+
+        $this->logger->notice('Received an authentication response for self vetting a second factor');
+
+        /** @var SelfVetRequestId $initiatedRequestId */
+        $initiatedRequestId = $this->session->get(self::SELF_VET_SESSION_ID);
+
+        $samlLogger = $this->samlLogger->forAuthentication($initiatedRequestId->requestId());
+
+        $this->session->remove(self::SELF_VET_SESSION_ID);
+
+        try {
+            $assertion = $this->postBinding->processResponse(
+                $httpRequest,
+                $this->identityProvider,
+                $this->serviceProvider
+            );
+
+            if (!InResponseTo::assertEquals($assertion, $initiatedRequestId->requestId())) {
+                $samlLogger->error(
+                    sprintf(
+                        'Expected a response to the request with ID "%s", but the SAMLResponse was a response to a different request',
+                        $initiatedRequestId
+                    )
+                );
+                throw new AuthenticationException('Unexpected InResponseTo in SAMLResponse');
+            }
+            $candidateSecondFactor = $this->secondFactorService->findOneVerified($secondFactorId);
+            // Proof of possession of higher/equal LoA was successful, now apply the self vet command on Middleware
+            $command = new SelfVetCommand();
+            $command->identity = $this->getIdentity();
+            $command->secondFactor = $candidateSecondFactor;
+            $command->authoringLoa = $assertion->getAuthnContextClassRef();
```

Comment:
Why not send the authorizing secondFactorID, so the LOA could be validated in MW because there is were the domain logic should be validated. Consider also sending the authoring LOA so MW could validate if the vetted token has the LOA as expected by MW.

Reply:
I do not know how to retrieve the token used in the gateway authentication. There is a possibility the users possesses multiple suitable SF tokens, and the WAYG is used to choose to token of his/her liking. The assertion only displays the LOa level that was reached, not the Second Factor ID. Unless maybe this can be fetched from a magic session I'm not aware of but you might?

Reply:
I was under the assumption that you would have to select the token you would authorize the token with in SA. That way you would have a decent audit trail and MW would be the SSOT. Do you know the motives why this wasn't done?

Reply:
That might have been discussed at some point, but re-using the existing WAYG in Gateway was decided on. 

1. Faster/easier to develop
2. Prevent code duplication

Reply:
Clear, but be aware that this will be hard to change later on due to the event sourced MW.